File: microfundamentos/AOP/2-Ordenacao/heap/Main.java
import java.io.IOException;
import java.util.Scanner;

class Heap {
    private static final int TAM = 9;
    int[] array;
    int tam;

    public Heap() {
        this.tam = 0;
        this.array = new int[TAM];
    }

    public int getTAM() {
        return TAM;
    }

    public void mostrarHeap() {
        if(tam == 0) {
            System.out.println("Heap vazia\n"); return;
        }

        System.out.println("Heap tam: "+ tam);
        for(int i = 0; i < tam; i++) System.out.print(array[i] + " "); System.out.println();
    }

    public void inserirHeap(int elem) {
        if(tam >= TAM) {
            System.out.println("Heap cheia"); return;
        }

        array[tam++] = elem;
    }
}

class Main {

    public static void main(String[] args) {

        Heap minhaHeap = new Heap();

        minhaHeap.mostrarHeap();

        minhaHeap.inserirHeap(5);
        minhaHeap.inserirHeap(2);
        minhaHeap.inserirHeap(7);

        minhaHeap.mostrarHeap();

        System.out.println(minhaHeap.getTAM());
    }
}

File: microfundamentos/AOP/2-Ordenacao/heap/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#define TAM 9

typedef struct {
    int array[TAM];
    int tam;
} Heap;

Heap* criarHeap() {
    Heap* h = (Heap*)malloc(sizeof(Heap));
    h->tam = 0;
    
    return h;
}

void mostrarHeap(Heap* h) {
    if(h->tam == 0) {
        printf("Heap vazia\n"); return;
    }

    printf("Heap tam: %d\n", h->tam);
    for(int i = 0; i < h->tam; i++) printf("%d ", h->array[i]); printf("\n");
}

void inserirHeap(Heap* h, int elem) {
    
    if(h->tam >= TAM) {
        printf("Heap cheia\n"); return;
    }
    
    h->array[h->tam++] = elem;
}


int main(void) {
    
    Heap* minhaheap = criarHeap();
    
    mostrarHeap(minhaheap);
    
    inserirHeap(minhaheap, 5);
    inserirHeap(minhaheap, 2);
    inserirHeap(minhaheap, 7);
    mostrarHeap(minhaheap);
    
    return 0;
}